### PR TITLE
[11.14] GraphQL API

### LINKED ISSUE #766: GraphQL API
Gitlab supports GraphQL: https://docs.gitlab.com/ee/api/graphql/

Github equivalent of this feature in the respective PHP SDK: https://github.com/KnpLabs/php-github-api/blob/master/lib/Github/Api/GraphQL.php

### PR DESCRIPTION
**Pull request by claudiu-cristea**
Fixes #766

### CODE REVIEW

File: src/Api/AbstractApi.php [hunk 1]
```diff
@@ -36,7 +36,7 @@ abstract class AbstractApi
      *
      * @var string
      */
-    private const URI_PREFIX = '/api/v4/';
+    protected const URI_PREFIX = '/api/v4/';
```
Comment (dkarlovi) [inline src/Api/AbstractApi.php:39]: Do you really need to change this? Since you're using static now, it should automatically use the one from the child class, no? 🤔

Comment (claudiu-cristea) [inline src/Api/AbstractApi.php:39]: Maybe we should split version out of `URI_PREFIX`, into a protected class? Then we override the version in GraphQL as GitLab GraphQL is versionless

File: src/Api/GraphQL.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ * (c) Matt Humphrey <matth@windsor-telecom.co.uk>
+ * (c) Graham Campbell <hello@gjcampbell.co.uk>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Gitlab\Api;
+
+class GraphQL extends AbstractApi
```
Comment (dkarlovi) [inline src/Api/GraphQL.php:17]: Should have a few test cases.

**Review by dkarlovi — CHANGES_REQUESTED** [review]

File: src/Api/AbstractApi.php (hunk 1 above)
Comment (claudiu-cristea) [reply src/Api/AbstractApi.php:39]: In PHP `private` means a sub-class cannot extend/override it.

File: src/Api/GraphQL.php (hunk 2 above)
Comment (claudiu-cristea) [reply src/Api/GraphQL.php:17]: We should. But I have no time to work on

File: src/Api/AbstractApi.php [hunk 3]
```diff
@@ -264,7 +264,7 @@ private static function prepareUri(string $uri, array $query = []): string
             return null !== $value;
         });
 
-        return \sprintf('%s%s%s', self::URI_PREFIX, $uri, QueryStringBuilder::build($query));
+        return \sprintf('%s%s%s', static::URI_PREFIX, $uri, QueryStringBuilder::build($query));
```
Comment (GrahamCampbell) [inline src/Api/AbstractApi.php:267]: Please revert these changes.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Api/AbstractApi.php (hunk 3 above)
Comment (claudiu-cristea) [reply src/Api/AbstractApi.php:267]: I don't think will work. GraphQL needs a different URL prefix

File: src/Api/AbstractApi.php (hunk 1 above)
Comment (GrahamCampbell) [reply src/Api/AbstractApi.php:39]: why does it need to be overwritten? the changes you want don't need to re-use this?

Comment (claudiu-cristea) [reply src/Api/AbstractApi.php:39]: @GrahamCampbell, GraphQL is _versionless_. The URL prefix is `/api` instead of `/api/v4`, so `AbstractApi::prepareUriP()`, needs to get `/api` when subclassed as `\Gitlab\Api\GraphQL`

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]
I think it's a mistake to extend `AbstractApi` at all.

**Comment by claudiu-cristea** [comment]
> I think it's a mistake to extend `AbstractApi` at all.

Maybe but I wanted to reuse `AbstractApi::post()`. Or should we move that in a trait?

**Comment by GrahamCampbell** [comment]
We should just take the hit on a little duplication. Traits are horrible for this type of thing (most things imo :trollface:).

**Comment by GrahamCampbell** [comment]
Closing due to inactivity.
